File: php/tests/Migrations/DerivedDownTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Migrations;

use DateTimeImmutable;
use PDO;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Dialect\SqliteDialect;
use SimpleOrm\Errors\SimpleOrmException;
use SimpleOrm\Metadata\EntityMapLoader;
use SimpleOrm\Migrations\ColumnRename;
use SimpleOrm\Migrations\MigrationRunner;
use SimpleOrm\Migrations\MigrationSet;
use SimpleOrm\Migrations\SchemaSnapshot;
use SimpleOrm\Migrations\SnapshotSet;
use SimpleOrm\Migrations\SqlVersion;
use SimpleOrm\Migrations\SqlVersionStep;
use SimpleOrm\Migrations\TableSchema;
use SimpleOrm\Migrations\TableSchemaColumn;
use SimpleOrm\Tests\Support\TempDatabase;

/**
 * ADR-0018 (owner): "Down could be deducted from the previous schema" — nobody
 * writes rollback DDL. Mirrors `dotnet/tests/SimpleOrm.Tests/DerivedDownTests.cs`
 * at the `MigrationRunner` level (the pure-`DownDeriver` unit tests live in
 * `DownDeriverTest`): the runner derives the rollback at migrate-down time from
 * the versioned snapshots, inverting the step's typed renames data-preservingly
 * and diffing the rest. `down()` remains the manual override; missing snapshots
 * still refuse (`MIG-020`).
 */
final class DerivedDownTest extends TestCase
{
    private TempDatabase $database;

    private SqliteDialect $dialect;

    private PDO $connection;

    private EntityMapLoader $maps;

    protected function setUp(): void
    {
        $this->database = TempDatabase::create();
        $this->dialect = new SqliteDialect();
        $this->connection = $this->dialect->createConnection($this->database->connectionString());
        $this->maps = new EntityMapLoader();
    }

    protected function tearDown(): void
    {
        unset($this->connection);
        $this->database->delete();
    }

    /** @return list<SqlVersion> V1 creates widgets(id, name) and seeds 'admin'; V2 adds note; V3 renames name -> full_name. */
    private function history(): array
    {
        return [
            new SqlVersion(1, new SqlVersionStep(
                'widgets',
                'create',
                [
                    'create table widgets (id INTEGER PRIMARY KEY, name TEXT NOT NULL) STRICT',
                    "insert into widgets (name) values ('admin')",
                ],
            )),
            new SqlVersion(2, new SqlVersionStep('widgets', 'add_note', ['alter table widgets add column note TEXT'])),
            new SqlVersion(3, new SqlVersionStep(
                'widgets',
                'rename_name',
                ['alter table widgets rename column name to full_name'],
                renames: [new ColumnRename('name', 'full_name')],
            )),
        ];
    }

    private function historySnapshots(): SnapshotSet
    {
        $idColumn = new TableSchemaColumn('id', 'INTEGER', false, key: true, generated: true);
        $documents = [
            SchemaSnapshot::export(
                new TableSchema('widgets', [$idColumn, new TableSchemaColumn('name', 'TEXT', false)], []),
                1,
                new DateTimeImmutable(),
            ),
            SchemaSnapshot::export(
                new TableSchema('widgets', [
                    $idColumn,
                    new TableSchemaColumn('name', 'TEXT', false),
                    new TableSchemaColumn('note', 'TEXT', true),
                ], []),
                2,
                new DateTimeImmutable(),
            ),
            SchemaSnapshot::export(
                new TableSchema('widgets', [
                    $idColumn,
                    new TableSchemaColumn('full_name', 'TEXT', false),
                    new TableSchemaColumn('note', 'TEXT', true),
                ], []),
                3,
                new DateTimeImmutable(),
            ),
        ];

        return new SnapshotSet($documents);
    }

    private function runner(SnapshotSet $snapshots): MigrationRunner
    {
        return new MigrationRunner(
            $this->connection,
            $this->dialect,
            $this->maps,
            MigrationSet::of(...$this->history()),
            $snapshots,
        );
    }

    #[Test]
    public function the_whole_history_rolls_back_and_reapplies_without_any_down_code(): void
    {
        $runner = $this->runner($this->historySnapshots());

        self::assertSame(3, $runner->migrate());

        // No step overrides down(); every rollback below is derived — the
        // rename reverses, the added column drops, the create reverts to a
        // drop table.
        self::assertSame(3, $runner->migrateDown(0));
        self::assertSame(0, $this->countTables());

        // And the same plan applies cleanly again: down really reached V0000.
        self::assertSame(3, $runner->migrate());
        self::assertSame('admin', $this->scalar('select full_name from widgets'));
    }

    #[Test]
    public function partial_rollback_reverses_the_rename_and_keeps_data(): void
    {
        $runner = $this->runner($this->historySnapshots());
        $runner->migrate();

        // Down past V3 (name -> full_name): the derived rollback inverts the
        // rename, so the V1 'admin' seed survives under the old column name.
        $runner->migrateDown(2);

        self::assertSame('admin', $this->scalar('select name from widgets'));
    }

    #[Test]
    public function underivable_constraint_is_noticed_not_guessed(): void
    {
        // History: V1 creates derived_widgets(id, label NOT NULL); V2 drops label.
        $versions = [
            new SqlVersion(1, new SqlVersionStep(
                'derived_widgets',
                'create',
                ['create table derived_widgets (id INTEGER PRIMARY KEY, label TEXT NOT NULL) STRICT'],
            )),
            new SqlVersion(2, new SqlVersionStep('derived_widgets', 'drop_label', ['alter table derived_widgets drop column label'])),
        ];

        $idColumn = new TableSchemaColumn('id', 'INTEGER', false, key: true, generated: true);
        $snapshots = new SnapshotSet([
            SchemaSnapshot::export(
                new TableSchema('derived_widgets', [$idColumn, new TableSchemaColumn('label', 'TEXT', false)], []),
                1,
                new DateTimeImmutable(),
            ),
            SchemaSnapshot::export(new TableSchema('derived_widgets', [$idColumn], []), 2, new DateTimeImmutable()),
        ]);

        $withSnapshots = new MigrationRunner($this->connection, $this->dialect, $this->maps, MigrationSet::of(...$versions), $snapshots);
        $withSnapshots->migrate();

        // The structure returns nullable; the NOT NULL constraint (and the data) can't derive.
        $notices = [];
        self::assertSame(1, $withSnapshots->migrateDown(1, false, function (string $notice) use (&$notices): void {
            $notices[] = $notice;
        }));
        self::assertTrue(self::anyContains($notices, 'label') && self::anyContains($notices, 'not derivable'));

        // Without snapshots the same plan still refuses honestly (MIG-020).
        $withSnapshots->migrate();
        $blind = new MigrationRunner($this->connection, $this->dialect, $this->maps, MigrationSet::of(...$versions));
        try {
            $blind->migrateDown(1);
            self::fail('expected MIG-020');
        } catch (SimpleOrmException $exception) {
            self::assertSame('MIG-020', $exception->errorCode);
        }
    }

    private function countTables(): int
    {
        return (int) $this->scalar(
            "select count(*) from sqlite_master where type in ('table', 'view') "
                . "and name not like 'sqlite_%' and name <> 'schema_version'",
        );
    }

    private function scalar(string $sql): mixed
    {
        return $this->connection->query($sql)->fetchColumn();
    }

    /** @param list<string> $items */
    private static function anyContains(array $items, string $needle): bool
    {
        foreach ($items as $item) {
            if (str_contains($item, $needle)) {
                return true;
            }
        }

        return false;
    }
}

File: go/orm/migrations/derived_down_test.go
package migrations_test

import (
	"context"
	"strings"
	"testing"
	"time"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/migrations"
	"github.com/kosmassp/SimpleOrm/go/orm/sqlite"
)

// ADR-0018 (owner): "Down could be deducted from the previous schema" —
// nobody writes rollback DDL. The runner derives it at migrate-down time from
// the versioned snapshots, inverting the step's typed renames
// data-preservingly and diffing the rest. Down() remains the manual override;
// missing snapshots still refuse (MIG-020). Mirrors DerivedDownTests.cs; the
// whole-sample-history and partial-rollback scenarios live in
// migration_runner_test.go's TestRunner_SampleTreeRoundTripsThroughZero.

func TestDerivedDown_UnderivableConstraintIsNoticedNotGuessed(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	// History: V1 creates derived_widgets(id, label NOT NULL); V2 drops label.
	v1Schema := &migrations.TableSchema{Name: "derived_widgets", Columns: []migrations.TableColumn{
		{Name: "id", StorageType: "INTEGER", Nullable: false, Key: true, Generated: true},
		{Name: "label", StorageType: "TEXT", Nullable: false},
	}}
	v2Schema := &migrations.TableSchema{Name: "derived_widgets", Columns: []migrations.TableColumn{
		{Name: "id", StorageType: "INTEGER", Nullable: false, Key: true, Generated: true},
	}}

	dir := t.TempDir()
	writeSnapshotFile(t, dir, "s1.schema.json", migrations.ExportSchema(v1Schema, 1, time.Now()))
	writeSnapshotFile(t, dir, "s2.schema.json", migrations.ExportSchema(v2Schema, 2, time.Now()))
	snapshots, err := migrations.SnapshotsFromDirectory(dir)
	if err != nil {
		t.Fatal(err)
	}

	set, err := migrations.NewSet(
		rawTableVersion(1, "derived_widgets", "create",
			[]string{"create table derived_widgets (id INTEGER PRIMARY KEY, label TEXT NOT NULL) STRICT"}, nil),
		rawTableVersion(2, "derived_widgets", "drop label",
			[]string{"alter table derived_widgets drop column label"}, nil),
	)
	if err != nil {
		t.Fatal(err)
	}

	runner := migrations.NewRunner(pool, sqlite.New(), maps, set, snapshots)
	if _, err := runner.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	// The structure returns nullable; the NOT NULL constraint (and data) can't derive.
	var notices []string
	reverted, err := runner.MigrateDown(ctx, 1, migrations.RunOptions{Notify: func(n string) { notices = append(notices, n) }})
	if err != nil {
		t.Fatal(err)
	}
	if reverted != 1 {
		t.Fatalf("expected 1 reverted version, got %d", reverted)
	}
	found := false
	for _, n := range notices {
		if strings.Contains(n, "label") && strings.Contains(n, "not derivable") {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected a notice naming the underivable constraint, got %v", notices)
	}

	// Without snapshots the same plan still refuses honestly (MIG-020).
	if _, err := runner.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}
	blind := migrations.NewRunner(pool, sqlite.New(), maps, set, nil)
	_, err = blind.MigrateDown(ctx, 1, migrations.RunOptions{})
	if core.CodeOf(err) != "MIG-020" {
		t.Fatalf("expected MIG-020 without snapshots, got %v", err)
	}
}

func TestDerivedDown_CreateRevertsToDrop(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	maps := metadata.NewLoader(nil)

	schema := &migrations.TableSchema{Name: "widgets", Columns: []migrations.TableColumn{
		{Name: "id", StorageType: "INTEGER", Nullable: false, Key: true, Generated: true},
	}}
	dir := t.TempDir()
	writeSnapshotFile(t, dir, "s1.schema.json", migrations.ExportSchema(schema, 1, time.Now()))
	snapshots, err := migrations.SnapshotsFromDirectory(dir)
	if err != nil {
		t.Fatal(err)
	}

	set, err := migrations.NewSet(rawTableVersion(
		1, "widgets", "create", []string{"create table widgets (id INTEGER PRIMARY KEY) STRICT"}, nil))
	if err != nil {
		t.Fatal(err)
	}
	runner := migrations.NewRunner(pool, sqlite.New(), maps, set, snapshots)
	if _, err := runner.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}
	if _, err := runner.MigrateDown(ctx, 0, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	var count int64
	if err := pool.QueryRowContext(ctx,
		"select count(name) from sqlite_master where type = 'table' and name = 'widgets'").Scan(&count); err != nil {
		t.Fatal(err)
	}
	if count != 0 {
		t.Fatal("expected the derived rollback to drop the table created at V1")
	}
}
